ucp/_libs/ucx_api.pyx

```diff
@@ -373,6 +380,67 @@ def _ucx_worker_handle_finalizer(
 
 
 IF CY_UCP_AM_SUPPORTED:
+    cdef void _am_recv_completed_callback(
+        void *request,
+        ucs_status_t status,
+        size_t length,
+        void *user_data
```

Comment:
`void *user_data` doesn't seems to be used?

Reply:
Yes, this is the data buffer where the message is received. This is part of the [ucp_am_recv_data_nbx_callback_t](http://openucx.github.io/ucx/api/latest/html/group___u_c_p___c_o_m_m.html#ga242399982563973351591516039dd74b) prototype, but the copy is handled automatically by UCX, so we don't need to do anything with it.

Reply:
Yeah, but you are assigning it in `_am_recv_callback()`:
```
            request_param.user_data = <void *>buf
            status = ucp_am_recv_data_nbx(
                worker._handle, data, buf_ptr, length, &request_param
            )
```
and checking that it isn't `NULL` in `_am_recv_completed_callback()`:
```
assert user_data != NULL
```

I know it is a minor thing but it is just a bit confusing :)

Reply:
Yes, I agree, it took me a while to understand that too. I see two options:

1. Don't assign `user_data` and lose the ability to log the data pointer; or
2. Keep `user_data` but remove the assertion, thus (hopefully) making it clearer it's only used for debugging.

I personally would prefer 2, do you have a preference here, or maybe even another suggestion on how to deal with this?

Reply:
>     1. Don't assign `user_data` and lose the ability to log the data pointer;

Isn't it possible to use `req_info["cb_args"]` for logging, which is also assigned the buffer pointer in `_am_recv_callback`?
```
req_info["cb_args"] = (buf, )
```
If not, maybe just remove `req_info["cb_args"]` and rely on `user_data` to get the buffer pointer?

Reply:
Yes, but that requires other checks before we can log anything, so if anything fails we lose the ability to do so anyway. I guess this is reasonable nevertheless, I'll do that.

Reply:
This is done now.